Fix docs.rs build and detect default Julia location

Darksonn opened the pull request:
The `dummy-bindings.rs` file is used instead of bindgen on docs.rs to avoid errors due to docs.rs not having julia installed. Additionally this will detect the default install location of Julia headers on most distros.

Darksonn committed:
Fix docs.rs build and detect default julia location

The dummy-bindings.rs is used instead of bindgen on docs.rs to avoid
errors due to docs.rs not having julia installed.

Taaitaaiger commented:
Thank you so much for this! I expect I'll review it this evening (CET).

Taaitaaiger commented on build.rs:
```diff
@@ -1,22 +1,62 @@
+    Path::new(path).exists()
+}
+
+fn find_julia() -> Option<String> {
```
The behaviour of this function is different depending on what branch it takes. If JL_PATH is set, it will just assume everything can be found through it. Otherwise it explicitly checks if julia.h exists in /usr/include/julia, failing that it performs the same explicit check for directories in either CPATH or C_INCLUDE_PATH, though it will only check one of the two. If it can't find julia.h, it panics at the end. While I do think it's good to check if /usr/include/julia/julia.h exists if JL_PATH is not set, I'd rather not have the additional (undocumented) checks for CPATH and C_INCLUDE_PATH and panic at the end.

Darksonn replied:
What do you wish to do if the `/usr/include/julia/julia.h` check fails?

Darksonn replied:
There's also the [pkg-config][1] crate which [this blog post][2] recommends, but I was not able to convince it to find Julia on my local machine, even though it was located at `/usr/include/julia/julia.h`.

[1]: https://crates.io/crates/pkg-config
[2]: https://kornel.ski/rust-sys-crate

Taaitaaiger replied:
I think it should just be accepted assuming "the user either ensures everything can be found through a default path, finds out they need to use JL_PATH, or they can open an issue if their use case is a hard requirement." My experience with installing Julia on Ubuntu is that the headers are not available, and I couldn't find a julia-dev package either, so I haven't tested it. If things are available the same way as they are in the official archives, your approach should work.

Darksonn replied:
What do you think about this?
```rust
fn find_julia() -> Option<String> {
    if let Ok(path) = env::var("JL_PATH") {
        return Some(path);
    }

    if path_exists("/usr/include/julia/julia.h") {
        return Some("/usr".to_string());
    }

    // This message is not displayed to the user unless the compile fails.
    eprintln!("You can specify the Julia installation path with the JL_PATH environment variable.");

    None
}
```
As for Ubuntu, I believe the package with the headers is called [`libjulia-dev`](https://packages.debian.org/unstable/libjulia-dev).

Taaitaaiger replied:
Looks great! I'll check if I can find libjulia-dev on Ubuntu 19.10 when I get home (I think I was always looking for julia-dev instead). In any case, it's not available on 18.04 as far as I'm aware (but neither is julia itself)

Taaitaaiger commented on build.rs:
```diff
@@ -1,22 +1,62 @@
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
+
+fn path_exists(path: &str) -> bool {
```
I don't think this needs to be a separate function.

Taaitaaiger commented on build.rs:
```diff
@@ -1,22 +1,62 @@
+    println!("cargo:rerun-if-changed=wrapper.h");
+    println!("cargo:rerun-if-env-changed=JL_PATH");
+
+    if let Ok(_) = env::var("CARGO_FEATURE_DOCS_RS") {
```
This works because of the stuff you've added to Cargo.toml, right? (I tested it by running docs.rs locally and it really seems to do the trick :)).

Darksonn replied:
Yep. The `docs-rs` feature is not a special feature name, but you can configure `docs.rs` to compile the crate with certain features enabled. You can also compile with
```
cargo doc --features docs-rs
```
to avoid having to run docs.rs locally.

Taaitaaiger replied:
Ah, TIL!

Taaitaaiger commented on dummy-bindings.rs:
```diff
@@ -0,0 +1,3677 @@
+/* automatically generated by rust-bindgen */
```
Please document what version of julia.h was used to generate these bindings in the module documentation of lib.rs, I think it's something that needs to be explicit to prevent confusion.

Darksonn replied:
It turns out to not be possible to embed the `//!` comment in the bindings file itself, but I have added a comment to the main `lib.rs` file:
```
//! The documentation found on docs.rs corresponds to Julia version 1.3.1, however when
//! compiled locally, the bindings will match the version installed locally.
```

Taaitaaiger replied:
That's what I was asking for :) That keeps this file completely generated

Taaitaaiger reviewed (changes requested):
There's a few small things I'd like you to change, would you mind doing that? In any case I've tested whether it works by running docs.rs in docker, and the dummy bindings really seem to do the trick. Thank you!

Darksonn committed:
Remove path_exists function

Darksonn committed:
Add Julia version to dummy bindings

Darksonn committed:
Try compiling even if Julia could not be found

Instead of a panic, just print a message that suggests setting the
JL_PATH environment variable if the compile fails.

Darksonn commented:
Will you release an updated version of the crate? It would be nice to be able to remove the `[patch]` section in my project's `Cargo.toml` and avoid having to set `JL_PATH` on every compile.

Taaitaaiger commented:
Sure! I've been doubting whether I do that now or wait until the next version of jlrs is ready, but it's the least I could do for your efforts. I'll comment again when I'm done

Taaitaaiger commented:
jl-sys 0.2.0 and jlrs 0.1.3 have been released! Thank you so much for making documentation of these crates available.
